### PR TITLE
Issue-3909: add ApolloResponse cache headers

### PR DESCRIPTION
**Pull request by ArjanSM**
see #[3909](https://github.com/apollographql/apollo-kotlin/issues/3909)

Suggestion Required.
Merging the `CacheHeaders` of `ApolloRequest` & `ApolloResponse`, does pose the risk of indeed overwriting the header values that have the same key. I am wondering if the team is okay with that?

### CODE REVIEW

**Comment by martinbonnin** [comment]
Thanks for sending this. Looks very good to me 👍 . We'll want to add test to document this and make sure the behaviour isn't changed moving forward. 
I was thinking an integration test, certainly in `CacheFlagsTest.ks`. Do you want to look into it? (integration tests are a composite build you can use by opening the `tests` folder in IntelliJ/Android Studio)

**Comment by ArjanSM** [comment]
> Thanks for sending this. Looks very good to me 👍 . We'll want to add test to document this and make sure the behaviour isn't changed moving forward. I was thinking an integration test, certainly in `CacheFlagsTest.ks`. Do you want to look into it? (integration tests are a composite build you can use by opening the `tests` folder in IntelliJ/Android Studio)

Thanks for the vote of confidence. 😄 
Sure! I can give it a try!

**Comment by martinbonnin** [comment]
Also you'll need to run `./gradlew apiDump` and commit the updated files to track the new "plus" API

**Comment by BoD** [comment]
Thanks for the PR!

> Merging the CacheHeaders of ApolloRequest & ApolloResponse, does pose the risk of indeed overwriting the header values that have the same key. I am wondering if the team is okay with that?

I think that's ok - the only risk I see is if interceptors set one of the headers (in `ApolloCacheHeaders`) that the cache expects, for some reason. I don't think it's very likely to happen 😊.

**Comment by ArjanSM** [comment]
> > Thanks for sending this. Looks very good to me 👍 . We'll want to add test to document this and make sure the behaviour isn't changed moving forward. I was thinking an integration test, certainly in `CacheFlagsTest.ks`. Do you want to look into it? (integration tests are a composite build you can use by opening the `tests` folder in IntelliJ/Android Studio)
> 
> Thanks for the vote of confidence. 😄 Sure! I can give it a try!

@martinbonnin Any suggestions on how to test it? Apparently, all my test cases are failing. Any suggestions will be welcome.

**Comment by martinbonnin** [comment]
The CI fails currently because of the API files not being updated, it should become all green once you do the `./gradlew ciDump` and commit the files.

As for testing, I'd recommend adding a test in [CacheFlagTest](https://github.com/apollographql/apollo-kotlin/blob/f476e024c0c74e222270cc357ee48202df43ee0a/tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt#L27). Ideally something that describes your use case so that it explains the motivation behind the change. Or if you want to send me some code, I can integrate it in the tests.

File: tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt [hunk 1]
```diff
@@ -101,4 +108,26 @@ class CacheFlagsTest {
...
+
+  @Test
+  fun doNotStoreWhenSetInResponse() = runTest(before = { setUp() }) {
+    val query = HeroNameQuery()
+    val data = HeroNameQuery.Data(HeroNameQuery.Hero("R2-D2"))
+    apolloClient = apolloClient.newBuilder().addInterceptor(object: ApolloInterceptor{
+      override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
+        return chain.proceed(request).onEach { response ->
+          val x = response.newBuilder().cacheHeaders(CacheHeaders.Builder().addHeader(ApolloCacheHeaders.DO_NOT_STORE, "").build()).build()
+          x
```
Comment (martinbonnin) [inline tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt:120]: Nit:
```suggestion
          response.newBuilder().cacheHeaders(CacheHeaders.Builder().addHeader(ApolloCacheHeaders.DO_NOT_STORE, "").build()).build()
```

File: tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt [hunk 2]
```diff
@@ -101,4 +108,26 @@ class CacheFlagsTest {
...
+        }
+      }
+    }).build()
+    apolloClient.enqueueTestResponse(query, data)
+
+    apolloClient.query(query).fetchPolicy(FetchPolicy.NetworkFirst).execute()
+
+    // Since the previous request was not stored, this should fail
+    assertFailsWith(CacheMissException::class) {
+      apolloClient.query(query).fetchPolicy(FetchPolicy.CacheOnly).execute()
```
[the remark this answers is not in the document]
Comment (ArjanSM) [reply tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt:130]: But for some weird reason this test keeps failing. :-(

[the remark this answers is not in the document]
Comment (martinbonnin) [reply tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt:129]: Interesting, let me check...

[the remark this answers is not in the document]
Comment (martinbonnin) [reply tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt:129]: Ah, you need to `map`, not `onEach`

File: tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt [hunk 3]
```diff
@@ -101,4 +108,25 @@ class CacheFlagsTest {
...
+
+  @Test
+  fun doNotStoreWhenSetInResponse() = runTest(before = { setUp() }) {
+    val query = HeroNameQuery()
+    val data = HeroNameQuery.Data(HeroNameQuery.Hero("R2-D2"))
+    apolloClient = apolloClient.newBuilder().addInterceptor(object: ApolloInterceptor{
+      override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
+        return chain.proceed(request).onEach { response ->
+          response.newBuilder().cacheHeaders(CacheHeaders.Builder().addHeader(ApolloCacheHeaders.DO_NOT_STORE, "").build()).build()
+        }
```
Comment (martinbonnin) [inline tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt:120]:
```suggestion
        return chain.proceed(request).map { response ->
          response.newBuilder().cacheHeaders(CacheHeaders.Builder().addHeader(ApolloCacheHeaders.DO_NOT_STORE, "").build()).build()
        }
```

File: tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArjanSM) [reply tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt:130]: Wow! TIL. Thank you so much @martinbonnin

[the remark this answers is not in the document]
Comment (martinbonnin) [reply tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt:130]: Sure thing!

File: tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt (hunk 3 above)
Comment (ArjanSM) [reply tests/integration-tests/src/commonTest/kotlin/test/CacheFlagsTest.kt]: Thanks!

**Review by martinbonnin — APPROVED** [review]

**Review by BoD — APPROVED** [review]

**Pull request merged**